### PR TITLE
Added Paging option to Delivery Pipeline

### PR DESCRIPTION
**Pull request by pskumar448**
Added ability to browser previous pipelines using Paging

Option provided to Enable/Disable Paging.
Initially Paging option will be false and DPP behaves like default design or behavior 

After Enabling Paging option through configure page, the noOfPipelines becomes the Page Size.
Provided Paging for component wise, at a time only one component can be browse through pipelines.

This gives us to browse old pipelines when they are exceeded 50 pipelines.

Fixes 
https://issues.jenkins-ci.org/browse/JENKINS-28918

### CODE REVIEW

**Comment by pskumar448** [comment]
@tommysdk 
Could you please look into this and provide your feedback regarding this feature.

File: src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java [hunk 1]
```diff
@@ -105,11 +106,13 @@
...
     private boolean showStaticAnalysisResults = false;
+    private boolean enablePaging = false;    
+    
 
     private List<RegExpSpec> regexpFirstJobs;
 
-    private transient String error;
-
+    private transient String error;    
+
```
Comment (szpak) [inline src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java]: Accidental reformat?

File: src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java [hunk 2]
```diff
@@ -436,6 +448,10 @@ protected static String withoutFolderPrefix(final String projectName) {
             }
             LOG.fine("Returning: " + components);
             error = null;
+            for(int i = 0; i< components.size(); i++)
+            {
```
Comment (szpak) [inline src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java]: `{` in the new line is not consistent with other places in code.

File: src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	
+	public PipelinePagination(int currentPage, int totalCount, int pagingSize, String moveLink) {
+		this.currentPage = currentPage;
+		this.totalCount = totalCount;
+		this.pageSize = pagingSize == 0 ? 10 : pagingSize;
+		this.moveLink = moveLink;
+		try {
+			rootUrl = Hudson.getInstance().getRootUrlFromRequest();
+		} catch (Exception e) {
+
```
Comment (szpak) [inline src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java]: It should be at least logged if is not crucial to make it work. 1. When can it fail? 2. Do you need to catch `Exception`?

File: src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	
+	public String getTag() {		
+		setPageNation();
+		StringBuilder sb = new StringBuilder();
+		//sb.append("<table summary='page navigation' class='Nnavi'>\n").append("<tr>\n");
+		if (prevPage + 1 > 10) {
+			setPrePage(sb);
+		} else {
+			//setBlankSpan(sb);
```
Comment (szpak) [inline src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java]: Why it is commented out?

File: src/main/java/se/diabol/jenkins/pipeline/domain/Component.java [hunk 5]
```diff
@@ -31,19 +36,34 @@
...
         this.firstJobUrl = firstJobUrl;
         this.firstJobParameterized = firstJobParameterized;
+        this.noOfPipelines = noOfPipelines;
+        this.enablePaging = enablePaging;
     }
 
     @Exported
     public List<Pipeline> getPipelines() {
-        return pipelines;
+    	if(enablePaging) {
```
Comment (szpak) [inline src/main/java/se/diabol/jenkins/pipeline/domain/Component.java]: Inconsistent formatting - missing a space between `if ` and `(`. In multiple places.

File: src/main/java/se/diabol/jenkins/pipeline/domain/Component.java [hunk 6]
```diff
@@ -65,4 +85,48 @@ public boolean isFirstJobParameterized() {
...
+    	
+    	return getPagination().getTag();    	
+    }
+    
+    public PipelinePagination getPagination() {    	
+    	if(enablePaging) {
+    		return new PipelinePagination(this.getCurrentPage(), pipelines.size(), noOfPipelines, "?" + (this.isFullScreenView() == true ? "fullscreen=true&" : "fullscreen=false&") + "component=" + componentNumber + "&page=");    	
+    	}
+    	return null;
+	}
```
Comment (szpak) [inline src/main/java/se/diabol/jenkins/pipeline/domain/Component.java]: Formatting again.

File: src/main/java/se/diabol/jenkins/pipeline/domain/Component.java [hunk 7]
```diff
@@ -65,4 +85,48 @@ public boolean isFirstJobParameterized() {
...
+    
+    public int getCurrentPage() {
+    	StaplerRequest req = Stapler.getCurrentRequest();    
+    	if(req == null)
```
Comment (szpak) [inline src/main/java/se/diabol/jenkins/pipeline/domain/Component.java]: Missing `{}` (and in a few other places).

**Comment by tommysdk** [comment]
Hi! I've gone through the pull request and I have some things I would like to have addressed. The default view pagination looks sufficient, although it doesn't make sense to have pagination on the fullscreen view, so that I would like to have removed all together. I'll update the JIRA issue to reflect that. Then there is a lot of inconsistencies in the source code regarding the style / formatting which I would have aligned with how the code is structured elsewhere.

File: src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java (hunk 1 above)
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java]: Please revert unnecessary reformat / addition of whitespace

File: src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java [hunk 8]
```diff
@@ -105,11 +106,13 @@
     private boolean showPromotions = false;
     private boolean showTestResults = false;
     private boolean showStaticAnalysisResults = false;
+    private boolean enablePaging = false;
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java]: Could you please rename it to pagingEnabled as I think it reads better?

File: src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java (hunk 4 above)
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java]: @pskumar448 could you please explain the intention of the commented out lines? We should not have any commented out code, if it isn't relevant, please just remove those lines.

File: src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+			} else {				
+				sb.append("<a href='").append(moveLink).append(count).append("'>").append(count).append("</a>\n");
+			}
+			
+		}
+
+	}
+
+	private void setPageNation() {
+		int current = (currentPage - 1) / 10 + 1;
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java]: I think the intendation of the file is wrong, there seem to be too many spaces added when the code is intended. We don't currently have any style guidelines I'm afraid but a good start is to use the standard Java convention

File: src/main/java/se/diabol/jenkins/pipeline/domain/Component.java [hunk 10]
```diff
@@ -31,19 +36,34 @@
...
     private final String firstJobUrl;
     private final boolean firstJobParameterized;
+    
+    private final int noOfPipelines;
+    
+    private int componentNumber = 0;
+    
+    private String pagingData = null;
+    
+    private boolean enablePaging = false;
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/domain/Component.java]: Same as in DeliveryPipelineView; Could you please rename it to pagingEnabled as it reads better?

File: src/main/java/se/diabol/jenkins/pipeline/domain/Component.java [hunk 11]
```diff
@@ -65,4 +85,48 @@ public boolean isFirstJobParameterized() {
     public String toString() {
         return toStringHelper(this).add("name", getName()).add("pipelines", pipelines).toString();
     }
+    
+    @Exported
+    public String getPagingData() {
+    	if(getPagination() == null)
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/domain/Component.java]: Please add braces around the if expression and reformat

File: src/main/java/se/diabol/jenkins/pipeline/domain/Pipeline.java [hunk 12]
```diff
@@ -222,10 +222,14 @@ public Pipeline createPipelineAggregated(ItemGroup context) {
             result.add(pipelineLatest);
             no--;
         }
-
-
+        
+        int pipelineCount = noOfPipelines;
+        if(enablePaging)
+        {
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/domain/Pipeline.java]: Please put on end of line 227

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly [hunk 13]
```diff
@@ -72,6 +72,10 @@
         <f:entry title="URL for custom CSS file (fullscreen)" field="fullScreenCss">
             <f:textbox/>
         </f:entry>
+        
+        <f:entry title="Enable Paging" field="enablePaging">
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly]: Please update to "Enable Paging" (lower-case p) to align with naming convention

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly [hunk 14]
```diff
@@ -49,7 +55,7 @@
     <div id="pipelineerror" class="pipelineerror"/>
 
     <div id="pipeline-message" class="pipeline-message"><div class="pipeline-loading-icon"></div></div>
-
+
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly]: Unnecessary whitespace?

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly [hunk 15]
```diff
@@ -72,8 +78,8 @@
         });
 
         view.viewUrl = '${it.getViewUrl()}';
-
-        updatePipelines(pipelineContainers, "pipelineerror", view, fullscreen, ${it.isShowChanges()},
+
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly]: Unnecessary whitespace?

File: src/test/java/se/diabol/jenkins/pipeline/domain/PipelineTest.java [hunk 16]
```diff
@@ -418,7 +418,7 @@ public void testFirstUpstreamBuildFirstProjectHasJustOneUpstreamJob() throws Exc
 
         assertEquals(build.getLastBuild(), BuildUtil.getFirstUpstreamBuild(build.getLastBuild(), build));
         Pipeline pipeline = Pipeline.extractPipeline("Pipeline", build);
-        List<Pipeline> pipelines = pipeline.createPipelineLatest(1, Jenkins.getInstance());
+        List<Pipeline> pipelines = pipeline.createPipelineLatest(1, Jenkins.getInstance(), false);
```
Comment (tommysdk) [inline src/test/java/se/diabol/jenkins/pipeline/domain/PipelineTest.java]: Please extract false to a constant declared in the beginning of the class to declare the intent of the parameter, which then can be reused throughout the test file

File: src/test/java/se/diabol/jenkins/pipeline/domain/status/SimpleStatusTest.java [hunk 17]
```diff
@@ -353,21 +353,21 @@ public void testCheckThatAllOnlyQueuedBuildIsResolvedAsQueued() throws Exception
         jenkins.buildAndAssertSuccess(project1);
         jenkins.waitUntilNoActivity();
 
-        List<Pipeline> pipelines = pipeline.createPipelineLatest(2, jenkins.getInstance());
+        List<Pipeline> pipelines = pipeline.createPipelineLatest(2, jenkins.getInstance(), false);
```
Comment (tommysdk) [inline src/test/java/se/diabol/jenkins/pipeline/domain/status/SimpleStatusTest.java]: extract false to a constant declared in the beginning of the class to declare the intent of the parameter, which then can be reused throughout the test file

File: src/test/java/se/diabol/jenkins/pipeline/sort/LatestActivityComparatorTest.java [hunk 18]
```diff
@@ -61,8 +61,8 @@ public void testCompare() {
         pipelinesB.add(pipelineB);
 
 
-        Component componentB = new Component("B", "B", "job/A", false, pipelinesB);
-        Component componentA = new Component("A", "A", "job/B", false, pipelinesA);
+        Component componentB = new Component("B", "B", "job/A", false, pipelinesB,3,false);
```
Comment (tommysdk) [inline src/test/java/se/diabol/jenkins/pipeline/sort/LatestActivityComparatorTest.java]: extract false to a constant declared in the beginning of the class to declare the intent of the parameter, which then can be reused throughout the test file

File: src/test/java/se/diabol/jenkins/pipeline/sort/NameComparatorTest.java [hunk 19]
```diff
@@ -31,8 +31,8 @@
 
     @Test
     public void testCompare() {
-        Component componentB = new Component("B", "B", "job/A", false, new ArrayList<Pipeline>());
-        Component componentA = new Component("A", "A", "job/B", false, new ArrayList<Pipeline>());
+        Component componentB = new Component("B", "B", "job/A", false, new ArrayList<Pipeline>(), 3, false);
```
Comment (tommysdk) [inline src/test/java/se/diabol/jenkins/pipeline/sort/NameComparatorTest.java]: extract false to a constant declared in the beginning of the class to declare the intent of the parameter, which then can be reused throughout the test file

**Comment by tommysdk** [comment]
Please also see the comments in the source code I added. Also, thank you @szpak for the feedback, much appreciated!

**Comment by pskumar448** [comment]
Ok @tommysdk 
I will take up the review points and get back to you once I finished.
Review Points: For summarization
1) Removing the pagination in full screen view.
2) Format & Styling issues needs to fixed or deleted.

Thanks @tommysdk , @szpak  for quick feedback.

**Comment by tommysdk** [comment]
@pskumar448 Great stuff! Looking forward to the improvements. Thank you!

File: src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java (hunk 4 above)
Comment (pskumar448) [inline src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java]: @szpak 
The reason for commented out those lines, initially I tried with another pagination style for that it required after that I shifted to another that I did commenting and forgot to remove the lines.
Now I am taking care of all the review feedback and working on it.

**Comment by pskumar448** [comment]
@tommysdk 
Done with following items
1) Removing the pagination in full screen view.
2) Format & Styling issues needs to fixed or deleted.

Please review and let me know the feedback.

File: src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java [hunk 20]
```diff
@@ -105,6 +106,7 @@
     private boolean showPromotions = false;
     private boolean showTestResults = false;
     private boolean showStaticAnalysisResults = false;
+    private boolean pagingEnabled = false;
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java]: neat!

File: src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+                sb.append(" class='active_link'>\n");
+                sb.append("<a>").append(count).append("</a>\n");
+                sb.append("</span>\n");
+            } else {
+                sb.append("<a href='").append(moveLink).append(count).append("'>").append(count).append("</a>\n");
+            }
+        }
+    }
+
+    private void setPageNation() {
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/PipelinePagination.java]: Shouldn't this be setPagination?

File: src/main/java/se/diabol/jenkins/pipeline/domain/Component.java [hunk 22]
```diff
@@ -31,18 +36,30 @@
...
         this.firstJobUrl = firstJobUrl;
         this.firstJobParameterized = firstJobParameterized;
+        this.noOfPipelines = noOfPipelines;
+        this.pagingEnabled = pagingEnabled;
     }
 
     @Exported
     public List<Pipeline> getPipelines() {
+        if (pagingEnabled) {
+            if (!isFullScreenView()) {
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/domain/Component.java]: Just collapse these two if statements
if (pagingEnabled && !isFullScreenView()) { ...

**Comment by tommysdk** [comment]
I think the feature looks good. Thanks for the code improvements. I added some comments on some minor things which would be nice to fix before merging.

I would also like you to rebase and squash the commits and amend the commit message with the issue number. Just prefix the commit message with JENKINS-28918.

If you're unfamiliar with squashing commits, here is an example: http://blog.diabol.se/?p=198 (section Squashing commits with Git).

When you've done that I think we can merge the feature.

**Comment by pskumar448** [comment]
@tommysdk 
Did rebase and squash. 
Provided comment message as suggested
 	JENKINS-28918 Added Paging option to Delivery Pipeline
Could you look into it.

**Comment by pskumar448** [comment]
@tommysdk 
I have done the Squash and Rebase, made as single commit.
Have a look and merge this feature.

**Comment by tommysdk** [comment]
Merged. Thanks for the contribution!

**Comment by pskumar448** [comment]
Thanks @tommysdk for guiding me by giving feedback.
